### PR TITLE
dynamo: dispatch `var_getattr` for `@lazy_property`

### LINKED ISSUE #93340: Dynamo doesn't support user defined decorator
### 🐛 Describe the bug

Repro
```
def fn(x):
    return torch.distributions.Categorical(probs=x).entropy()

opt_fn = torch._dynamo.optimize("eager")(fn)
x = torch.rand([4, 4])
print(opt_fn(x))
```
Error
```
  File "/scratch/ybliang/work/repos/pytorch/torch/_dynamo/symbolic_convert.py", line 990, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/scratch/ybliang/work/repos/pytorch/torch/_dynamo/symbolic_convert.py", line 461, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/scratch/ybliang/work/repos/pytorch/torch/_dynamo/variables/torch.py", line 205, in call_function
    self.as_python_constant()(
TypeError: finfo(): argument 'type' (position 1) must be torch.dtype, not NoneType

from user code:
   File "/scratch/ybliang/work/repos/pytorch/debug/debug1.py", line 23, in fn
    return torch.distributions.Categorical(probs=x).entropy()
  File "/scratch/ybliang/work/repos/pytorch/torch/distributions/categorical.py", line 130, in entropy
    min_real = torch.finfo(self.logits.dtype).min
```
This is because ```self.logits``` is decorated by ```lazy_property``` and Dynamo doesn't support generic decorator well.
https://github.com/pytorch/pytorch/blob/91a4947e2838be0420cc920ad35968f742dab24b/torch/distributions/categorical.py#L90-L92

I think there are two options to consider:
* If decorator is a common pattern used when writing neural network(personally I doubt), we should figure out a way to support it.
* Otherwise, maybe just special handle the ```Categorical.logits``` and similar attribute.

### Versions

N/A

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/93340

`lazy_property` is used in a lot of distributions, so I think it warrants special handling.

TODO:
- [ ] will add test 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @mlazos @soumith @yanboliang @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -326,6 +326,10 @@ def var_getattr(self, tx, name):
             )
         elif isinstance(subobj, types.FunctionType):
             return variables.UserMethodVariable(subobj, self, source=source, **options)
+        elif hasattr(subobj, "__get__"):
```
Comment (jon-chuang) [inline torch/_dynamo/variables/user_defined.py:329]: Alternately, we could dispatch on `@isinstance(.., lazy_property)`, it's definitely safer, I think this is not sound

Comment (jon-chuang) [inline torch/_dynamo/variables/user_defined.py:329]: Alternately, we could dispatch on `@isinstance(.., lazy_property)`, it's definitely safer, I think this is not sound

Comment (jon-chuang) [inline torch/_dynamo/variables/user_defined.py:329]: Alternately, we could dispatch on `@isinstance(.., lazy_property)`, it's definitely safer, I think this is not sound

Comment (jon-chuang) [reply torch/_dynamo/variables/user_defined.py:329]: Turns out this is indeed too permissive. I'll roll back to just @lazy_property

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -326,6 +326,10 @@ def var_getattr(self, tx, name):
             )
         elif isinstance(subobj, types.FunctionType):
             return variables.UserMethodVariable(subobj, self, source=source, **options)
+        elif isinstance(subobj, torch.distributions.utils.lazy_property):
+            return variables.UserMethodVariable(
+                subobj.__get__, self, source=source, **options
+            ).call_function(tx, [], {})
```
Comment (yanboliang) [inline torch/_dynamo/variables/user_defined.py:332]: I'm fine with this special handling, can you add a unit test?

**Comment by yanboliang** [comment]
Gently ping @jon-chuang!

**Pull request closed without merging**